storm-core/src/jvm/backtype/storm/utils/TransferDrainer.java

```diff
@@ -26,88 +26,47 @@
 
 public class TransferDrainer {
 
-  private HashMap<String, ArrayList<ArrayList<TaskMessage>>> bundles = new HashMap();
-  
-  public void add(HashMap<String, ArrayList<TaskMessage>> workerTupleSetMap) {
-    for (String key : workerTupleSetMap.keySet()) {
-      
-      ArrayList<ArrayList<TaskMessage>> bundle = bundles.get(key);
-      if (null == bundle) {
-        bundle = new ArrayList<ArrayList<TaskMessage>>();
-        bundles.put(key, bundle);
-      }
-      
-      ArrayList tupleSet = workerTupleSetMap.get(key);
-      if (null != tupleSet && tupleSet.size() > 0) {
-        bundle.add(tupleSet);
-      }
-    } 
+  private ArrayList<TaskMessage> buffer = new ArrayList<TaskMessage>();
+
+  public void addAll(ArrayList<TaskMessage> tuples) {
+    buffer.addAll(tuples);
```

Comment:
Yeah, I liked your suggestion of having the new buffer.  Before TransferDrainer was added, the buffer was just an ArrayList anyway.  I do not think it makes sense to have an ArrayList "inside" the TransferDrainer, because then we are implementing wrapper methods for `addAll` and `clear` that do not add value.